rules/register_toolchain.bzl

```diff
@@ -0,0 +1,238 @@
+        "runtime_classpath": attr.label_list(
+            mandatory = True,
+            providers = [JavaInfo],
+        ),
+        "semanticdb_bundle": attr.bool(
+            default = True,
+            doc = "Whether to bundle SemanticDB files in the resulting JAR. Note that in Scala 2, this requires the SemanticDB compiler plugin.",
+        ),
+        "use_ijar": attr.bool(
+            default = True,
+            doc = "Whether to use ijars for this compiler.",
+        ),
+        "version": attr.string(mandatory = True),
+        "_code_coverage_instrumentation_worker": attr.label(
+            allow_files = True,
+            cfg = "host",
+            default = "@rules_scala_annex//src/main/scala/higherkindness/rules_scala/workers/jacoco/instrumenter",
+            executable = True,
+        ),
+        "_compile_worker": attr.label(
+            allow_files = True,
+            cfg = "host",
+            default = "@rules_scala_annex//src/main/scala/higherkindness/rules_scala/workers/zinc/compile",
+            executable = True,
+        ),
+        "_deps_worker": attr.label(
+            allow_files = True,
+            cfg = "host",
+            default = "@rules_scala_annex//src/main/scala/higherkindness/rules_scala/workers/deps",
+            executable = True,
+        ),
+    },
+    implementation = _zinc_configuration_impl,
+)
+
+def _zinc_configuration(**kwargs):
+    if "deps_direct" not in kwargs:
+        kwargs["deps_direct"] = select({
+            "@rules_scala_annex//src/main/scala:deps_direct_off": "off",
+            "//conditions:default": "error",
+        })
+
+    if "deps_used" not in kwargs:
+        kwargs["deps_used"] = select({
+            "@rules_scala_annex//src/main/scala:deps_used_off": "off",
+            "//conditions:default": "error",
+        })
+
+    _zinc_configuration_underlying(**kwargs)
+
+def _make_register_toolchain(configuration_rule):
+    def result(name, visibility = ["//visibility:public"], **kwargs):
+        configuration_rule(
+            name = "{}-configuration".format(name),
+            visibility = visibility,
+            **kwargs
+        )
+
+        native.config_setting(
+            name = "{}-setting".format(name),
+            flag_values = {
+                "@rules_scala_annex_scala_toolchain//:scala-toolchain": name,
+            },
+        )
+
+        native.toolchain(
+            name = name,
+            target_settings = [":{}-setting".format(name)],
+            toolchain = ":{}-configuration".format(name),
+            toolchain_type = "@rules_scala_annex//rules/scala:toolchain_type",
+            visibility = visibility,
+        )
+
+    return result
+
+register_bootstrap_toolchain = _make_register_toolchain(_bootstrap_configuration)
+register_zinc_toolchain = _make_register_toolchain(_zinc_configuration)
+
+def _scala_toolchain_incoming_transition_impl(settings, attr):
+    if attr.scala_toolchain_name == "":
+        return {}
+
+    return {
```

Comment:
Do we want to be storing the original scala_toolchain_name like the `rules_go` folks did?

Because currently this will always reset to the original value, and I thought that you ran into a scenario where you could override things from the command line, which made that problematic.

I did this in the Play Routes and Twirl rules, in case it is at all helpful
```python
    new_settings = {}
    new_settings[toolchain_setting_key] = attr.play_routes_toolchain_name

    # Store the original toolchain value, so we can reset it
    new_settings[original_toolchain_setting_key] = json.encode(settings[toolchain_setting_key])

    return new_settings
```

Then I reset it as follows:
```python
    if settings[original_toolchain_setting_key] != "":
        new_settings = {}
        new_settings[toolchain_setting_key] = json.decode(settings[original_toolchain_setting_key])
        new_settings[original_toolchain_setting_key] = ""
        return new_settings
    else:
        return {}
```

Reply:
Ah, I completely forgot about that. I'll go ahead and do that.